File: lib/som/ruby/lib/distance.rb
class RubySOM
  module Distance
    
    def index_distance(closer, range)
      [ (closer[0]-range[0]).abs, (closer[1]-range[1]).abs ].max
    end
    
    def euclidean_distance(x, y)
      dist = 0
      x.each_index do |k|
        dist += (x[k]-y[k])**2
      end
      Math.sqrt(dist)
    end
    
    def closer(x, array)
      closer = [0, 0]
      array.each_index do |i|
        array[i].each_index do |j|
          p = [i, j]
          if array[i][j]==x
            closer = p
            break
          elsif euclidean_distance(x, array[p[0]][p[1]]) < euclidean_distance(x, array[closer[0]][closer[1]])
            closer = p
          end
        end
      end
      closer
    end
    
    def in_range?(x, closer, range)
      index_distance(closer, x) <= range
    end

  end
end

File: lib/som/java/src/neuralize/me/Distance.java
package neuralize.me;

import java.util.Arrays;
import java.util.List;

public class Distance {
	public static double euclideanDistance(List<Double> x, List<Double> y){
		double dist = 0;
		for(int i=0; i<x.size(); i++)
			dist += Math.pow(x.get(i) - y.get(i), 2); 
		return Math.sqrt(dist);
	}
	
	public static int indexDistance(List<Integer> closer, List<Integer> point){
		return Math.max( Math.abs(closer.get(0) - point.get(0)), Math.abs(closer.get(1)-point.get(1)) );
	}
	
	public static List<Integer> closer(List<Double> x, List<List<List<Double>>> array){
		List<Integer> closer = Arrays.asList(0, 0);
		for(int i=0; i< array.size(); i++){
			for(int j=0; j<array.get(i).size(); j++){
				List<Integer> point = Arrays.asList(i, j);
				
				if( array.get(i).get(j) == x ){
					closer = point;
					break;
				} else if( euclideanDistance(x, array.get(i).get(j)) < euclideanDistance(x, array.get(closer.get(0)).get(closer.get(1))) ) {
					closer = point;
				}
			}
		}
		
		return closer;
	}
	
	public static boolean inRange(List<Integer> point, List<Integer> closer, Integer range){
		return indexDistance(closer, point) <= range;
	}
}
